[RFC] Separate CPU offload activation to its own wrapper

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85495
* #86343
* #86337
* #86336
* #86331
* __->__ #85459

Passing in `offload_to_cpu=True` to checkpoint_wrapper is a bit confusing, because this causes the activation checkpoint args to be ignored and we do CPU offloading. This isn't ideal from API design perspective, so proposing to make `offload_wrapper` its own concept.

Now, offload to CPU + checkpoint can be composed together, such as

```
# apply AC to transformer layers
apply_ac_wrapper(model, checkpoint_wrapper, check_fn=lambda mod: isinstance(mod, TransformerLayer))
# offload the rest of activations to CPU
model = offload_wrapper(model)
```

Will polish / add tests if this proposal sounds good.

Differential Revision: [D39719854](https://our.internmc.facebook.com/intern/diff/D39719854/)

rohan-varma committed:
[RFC] Separate CPU offload activation to its own wrapper

Passing in `offload_to_cpu=True` to checkpoint_wrapper is a bit confusing, because this causes the activation checkpoint args to be ignored and we do CPU offloading. This isn't ideal from API design perspective, so proposing to make `offload_wrapper` its own concept.

Now, offload to CPU + checkpoint can be composed together, such as

```
# apply AC to transformer layers
apply_ac_wrapper(model, checkpoint_wrapper, check_fn=lambda mod: isinstance(mod, TransformerLayer))
# offload the rest of activations to CPU
model = offload_wrapper(model)
```

Will polish / add tests if this proposal sounds good.

Differential Revision: [D39719854](https://our.internmc.facebook.com/intern/diff/D39719854/)

[ghstack-poisoned]

vadimkantorov commented:
Related: https://github.com/pytorch/pytorch/issues/70166

Also, if autocast also existed as explicit wrapper, these problems would be easier to understand: https://github.com/pytorch/pytorch/issues/85340 https://github.com/pytorch/pytorch/issues/45910

rohan-varma committed:
Update on "[RFC] Separate CPU offload activation to its own wrapper"


Passing in `offload_to_cpu=True` to checkpoint_wrapper is a bit confusing, because this causes the activation checkpoint args to be ignored and we do CPU offloading. This isn't ideal from API design perspective, so proposing to make `offload_wrapper` its own concept.

Now, offload to CPU + checkpoint can be composed together, such as

```
# apply AC to transformer layers
apply_ac_wrapper(model, checkpoint_wrapper, check_fn=lambda mod: isinstance(mod, TransformerLayer))
# offload the rest of activations to CPU
model = offload_wrapper(model)
```

Will polish / add tests if this proposal sounds good.

Differential Revision: [D39719854](https://our.internmc.facebook.com/intern/diff/D39719854/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RFC] Separate CPU offload activation to its own wrapper"


Passing in `offload_to_cpu=True` to checkpoint_wrapper is a bit confusing, because this causes the activation checkpoint args to be ignored and we do CPU offloading. This isn't ideal from API design perspective, so proposing to make `offload_wrapper` its own concept.

Now, offload to CPU + checkpoint can be composed together, such as

```
# apply AC to transformer layers
apply_ac_wrapper(model, checkpoint_wrapper, check_fn=lambda mod: isinstance(mod, TransformerLayer))
# offload the rest of activations to CPU
model = offload_wrapper(model)
```

Will polish / add tests if this proposal sounds good.

Differential Revision: [D39719854](https://our.internmc.facebook.com/intern/diff/D39719854/)

[ghstack-poisoned]

awgu commented on torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:
```diff
@@ -219,13 +245,16 @@ def apply_activation_checkpointing(
+        # Checkpoint activations
         apply_activation_checkpointing(model, checkpoint_wrapper_fn=checkpoint_wrapper, check_fn=check_fn)
+        # Or Offload activations to CPU
+        apply_activation_checkpointing(model, checkpoint_wrapper_fn=offload_wrapper, check_fn=check_fn)
```
I think having the name `apply_activation_checkpointing()` as the entry point for both `checkpoint_wrapper` and `offload_wrapper` may be confusing since it suggests only the former.

It may be better to rename it to something like `apply_activation_wrapper()` to make it clear that it can be used for both activation _checkpointing_ and _offloading_ and it is done so via a module wrapper. (We can decide on a different name as well.)

awgu commented on torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:
```diff
@@ -155,10 +92,107 @@ def _pre_load_state_dict_hook(
+        super().__init__(mod)
+
+    def forward(self, *args, **kwargs):
+        with save_on_cpu(pin_memory=True):
```
Tangentially related: 

> You should not over-allocate pinned memory. Doing so can reduce overall system performance because it reduces the amount of physical memory available to the operating system and other programs. How much is too much is difficult to tell in advance, so as with all optimizations, test your applications and the systems they run on for optimal performance parameters.

https://developer.nvidia.com/blog/how-optimize-data-transfers-cuda-cc/

For activation offload, we operate under the assumption that we have unlimited memory to pin. I think it is worthwhile to investigate the robustness of this assumption.

Without any offloading, the total activation size is upper bounded by the GPU memory size, but with offloading, that can be increased significantly (e.g. <= the number of activations times the GPU memory size). For some workloads that activation offloading targets (e.g. large batch size and activations; small model), I imagine that the total activation size can be larger than the GPU memory size. This is exacerbated if CPU offloading is simultaneously enabled.

That being said, if the CPU memory size is sufficiently large and not used for much else during program execution, then pinning significant amounts of CPU memory may not be a problem. I think we should simply try to get some better intuition here.

rohan-varma replied:
This makes sense. I filed an issue: https://github.com/pytorch/pytorch/issues/86097. In the medium term if we do determine we shouldn't always pin memory, we might make this flag configurable or in the long-term possibly something that can be autotuned.

awgu commented on torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:
```diff
@@ -60,6 +32,9 @@ def __init__(
         )
 
+    def forward(self, *args, **kwargs):
+        raise ValueError("Implement me")
```
nit:
```suggestion
        raise NotImplementedError("Subclasses should implement this")
```

awgu commented on torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:
```diff
@@ -155,10 +92,107 @@ def _pre_load_state_dict_hook(
+
+class CheckpointWrapper(ActivationWrapper):
+    """
+    An nn.Module that wraps another nn.Module with checkpointing. Note that this
```
nit: Do we want these in code type?
```suggestion
    An ``nn.Module`` that wraps another ``nn.Module`` with checkpointing. Note that this
```

awgu commented on torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:
```diff
@@ -155,10 +92,107 @@ def _pre_load_state_dict_hook(
+        *checkpoint_fn_args,
+        **checkpoint_fn_kwargs,
+    ):
+        super().__init__()
```
The lint error seems real?
```suggestion
        super().__init__(mod)
```

awgu commented on torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:
```diff
@@ -155,10 +92,107 @@ def _pre_load_state_dict_hook(
+    offload can be composed with ones that do recomputation-based
+    checkpoint to trade off increased compute versus increased CPU
+    memory usage and additional H2D transfers.
+    Usage::
```
It looks like docs do not render for this file. If/when we do make the docs public, we should double check the docstring renders as expected.

awgu reviewed (commented):
Left some nits.

I think renaming to `apply_activation_wrapper()` in the short term is still reasonable. That being said, I am curious:
1. What is the cost to renaming it now? The previous PR renamed it, so it seems to be okay to rename.
2. What are the relevant future plans/discussions that may affect the naming?

rohan-varma committed:
Update on "[RFC] Separate CPU offload activation to its own wrapper"


Passing in `offload_to_cpu=True` to checkpoint_wrapper is a bit confusing, because this causes the activation checkpoint args to be ignored and we do CPU offloading. This isn't ideal from API design perspective, so proposing to make `offload_wrapper` its own concept.

Now, offload to CPU + checkpoint can be composed together, such as

```
# apply AC to transformer layers
apply_ac_wrapper(model, checkpoint_wrapper, check_fn=lambda mod: isinstance(mod, TransformerLayer))
# offload the rest of activations to CPU
model = offload_wrapper(model)
```

Will polish / add tests if this proposal sounds good.

Differential Revision: [D39719854](https://our.internmc.facebook.com/intern/diff/D39719854/)

[ghstack-poisoned]

rohan-varma commented:
discussed with @awgu to punt the renaming to the future when API is more finalized and functional approach is being taken.

rohan-varma committed:
Update on "[RFC] Separate CPU offload activation to its own wrapper"


Passing in `offload_to_cpu=True` to checkpoint_wrapper is a bit confusing, because this causes the activation checkpoint args to be ignored and we do CPU offloading. This isn't ideal from API design perspective, so proposing to make `offload_wrapper` its own concept.

Now, offload to CPU + checkpoint can be composed together, such as

```
# apply AC to transformer layers
apply_ac_wrapper(model, checkpoint_wrapper, check_fn=lambda mod: isinstance(mod, TransformerLayer))
# offload the rest of activations to CPU
model = offload_wrapper(model)
```

Will polish / add tests if this proposal sounds good.

Differential Revision: [D39719854](https://our.internmc.facebook.com/intern/diff/D39719854/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RFC] Separate CPU offload activation to its own wrapper"


Passing in `offload_to_cpu=True` to checkpoint_wrapper is a bit confusing, because this causes the activation checkpoint args to be ignored and we do CPU offloading. This isn't ideal from API design perspective, so proposing to make `offload_wrapper` its own concept.

Now, offload to CPU + checkpoint can be composed together, such as

```
# apply AC to transformer layers
apply_ac_wrapper(model, checkpoint_wrapper, check_fn=lambda mod: isinstance(mod, TransformerLayer))
# offload the rest of activations to CPU
model = offload_wrapper(model)
```

Will polish / add tests if this proposal sounds good.

Differential Revision: [D39719854](https://our.internmc.facebook.com/intern/diff/D39719854/)

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_checkpoint.py:
```diff
@@ -65,9 +66,11 @@ def __init__(
+                    ckpt_wrapper = offload_wrapper
+                else:
+                    ckpt_wrapper = checkpoint_wrapper
+                    pass
```
nit: remove the `pass`?

awgu commented on test/distributed/fsdp/test_fsdp_checkpoint.py:
```diff
@@ -166,13 +171,15 @@ def test_basic_checkpoint_end_to_end(self, cpu_offload, offload_activations):
             fsdp_only_seq = FSDP(deepcopy(seq), cpu_offload=cpu_offload)
             # Runs checkpoint-wrapped FSDP
+            if offload_activations:
+                checkpoint_wrapper = offload_wrapper
```
nit: maybe this should just be called `activation_wrapper` or `wrapper` (and change the instances below) or else you are shadowing the actual `checkpoint_wrapper` method, which might be confusing since it may be both `checkpoint_wrapper` or `offload_wrapper`

awgu reviewed (approved):
LGTM!

rohan-varma committed:
Update on "[RFC] Separate CPU offload activation to its own wrapper"


Passing in `offload_to_cpu=True` to checkpoint_wrapper is a bit confusing, because this causes the activation checkpoint args to be ignored and we do CPU offloading. This isn't ideal from API design perspective, so proposing to make `offload_wrapper` its own concept.

Now, offload to CPU + checkpoint can be composed together, such as

```
# apply AC to transformer layers
apply_ac_wrapper(model, checkpoint_wrapper, check_fn=lambda mod: isinstance(mod, TransformerLayer))
# offload the rest of activations to CPU
model = offload_wrapper(model)
```

Will polish / add tests if this proposal sounds good.

Differential Revision: [D39719854](https://our.internmc.facebook.com/intern/diff/D39719854/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RFC] Separate CPU offload activation to its own wrapper"


Passing in `offload_to_cpu=True` to checkpoint_wrapper is a bit confusing, because this causes the activation checkpoint args to be ignored and we do CPU offloading. This isn't ideal from API design perspective, so proposing to make `offload_wrapper` its own concept.

Now, offload to CPU + checkpoint can be composed together, such as

```
# apply AC to transformer layers
apply_ac_wrapper(model, checkpoint_wrapper, check_fn=lambda mod: isinstance(mod, TransformerLayer))
# offload the rest of activations to CPU
model = offload_wrapper(model)
```

Will polish / add tests if this proposal sounds good.

Differential Revision: [D39719854](https://our.internmc.facebook.com/intern/diff/D39719854/)

[ghstack-poisoned]

rohan-varma commented:
@rohan-varma has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39719854).

rohan-varma commented:
upload test status - Blocked is the failure, unrelated. Landing

rohan-varma commented:
@pytorchbot merge -f "CI failure unrelated"
